fq.go

```diff
@@ -0,0 +1,133 @@
+package fq
+
+import (
+	"math"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/util/clock"
+)
+
+type FQScheduler struct {
+	lock   sync.Mutex
+	queues []*Queue
+	clock  clock.Clock
+	// Verify float64 has enough bits
+	// We will want to be careful about having enough bits. For example, in float64, 1e20 + 1e0 == 1e20.
+	vt           float64
+	C            float64
+	G            float64
+	lastrealtime float64
+	robinidx     int
+}
+
+// TODO(aaron-prindle) add concurrency enforcement - 'C'
+func (q *FQScheduler) chooseQueue(packet *Packet) *Queue {
+	if packet.queueidx < 0 || packet.queueidx > len(q.queues) {
+		panic("no matching queue for packet")
+	}
+	return q.queues[packet.queueidx]
+}
+
+func NewFQScheduler(queues []*Queue, clock clock.Clock) *FQScheduler {
+	fq := &FQScheduler{
+		lock:   sync.Mutex{},
+		queues: queues,
+		clock:  clock,
+		vt:     0,
+	}
+	return fq
+}
+
+// TODO(aaron-prindle) verify that the time units are correct/matching
+func (q *FQScheduler) nowAsUnixNano() float64 {
+	return float64(q.clock.Now().UnixNano())
+}
+
+func (q *FQScheduler) Enqueue(packet *Packet) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+	q.synctime()
+
+	queue := q.chooseQueue(packet)
+	packet.starttime = q.nowAsUnixNano()
+
+	queue.enqueue(packet)
+	q.updateTime(packet, queue)
+}
+
+func (q *FQScheduler) now() float64 {
+	return q.vt
+}
+
+func (q *FQScheduler) synctime() {
+	now := q.nowAsUnixNano()
+	timesincelast := now - q.lastrealtime
```

Comment:
I updated this so that lastRealTime is a time.Time.  Is this suggesting to use Seconds vs Nanoseconds? I'm not sure I understand the reasoning:
> If we're using representing virtual time in floating point then there is no need to use tiny units.

Currently I have opted to use NanoSeconds() here but can change it to Seconds().  It seems to me that we would want to use the smallest time segment so that our virtual time has higher resolution.

Reply:
For a given `d Duration`, `d.Seconds()` --- which is a floating-point number, not an integer --- has as much resolution as `d.Nanoseconds()`, provided that the duration is no more than 104 days.  (An IEEE double-precision binary floating point number has 53 significand bits, and 2^53/10^9/24/60/60 is about 104.25).

Reply:
ah thanks, I did not realise d.Seconds() is a floating-point number. Updated to use Seconds()